protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,230 @@
+given the current CometBFT consensus algorithm
+(called Tendermint, and described in the [arXiv paper][arxiv]).
+We reproduce a slightly modified version of these requirements here.
+This version of the requirements is equivalent to the ones provided in the ABCI 1.0 specification.
+The modifications are needed in order to express other requirements later on in this document.
+
+Let $p$ and $q$ be two correct processes.
+Let $r_p$ be a round of consensus at height $h$ where $p$ is the proposer.
+Let $s_{p,h-1}$ be $p$'s application's state committed for height $h-1$.
+Let $v_p$ be the block that $p$'s CometBFT passes
+on to the application
+via `RequestPrepareProposal` as proposer of round $r_p$, height $h$,
+known as the _raw proposal_.
+Let $u_p$ be the (possibly modified) block that $p$'s application
+returns via `ResponsePrepareProposal` to CometBFT in round $r_p$, height $h$,
+known as the _prepared proposal_.
+
+These are the relevant requirements in the specification of ABCI 1.0 (and 2.0):
+
+* Requirement 3 [`PrepareProposal`, `ProcessProposal`, coherence]: For any two correct processes $p$ and $q$
+  and any round $r_p \geq 0$,
+  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
+  $q$'s application returns Accept in `ResponseProcessProposal`.
+
+* Requirement 4 [`ProcessProposal`, determinism-1]:
+  `ProcessProposal` is a (deterministic) function of the current
+  state and the block being processed.
+  In other words, for any correct process $p$, and any arbitrary block $u$,
+  if $p$'s CometBFT calls `RequestProcessProposal` on $u$ at height $h$,
+  then $p$'s application's acceptance or rejection **exclusively** depends on $u$ and $s_{p,h-1}$.
+
+* Requirement 5 [`ProcessProposal`, determinism-2]:
+  For any two correct processes *p* and *q*, and any arbitrary
+  block $u$,
+  if $p$'s (resp. $q$'s) CometBFT calls `RequestProcessProposal` on *u* at height $h$,
+  then $p$'s application accepts *u* if and only if $q$'s application accepts $u$.
+  Note that this requirement follows from the previous one and consensus _agreement_.
+
+The requirements expressed above are good enough for most applications using ABCI 1.0 or 2.0.
+They are simple to understand and it is relatively easy to check whether an application's
+implementation of `PrepareProposal` and `ProcessProposal` fulfills them.
+All applications that are able to enforce these properties do not need to reason about
+the internals of the consensus implementation: they can consider it as a black box.
+This is the most desirable situation in terms of modularity between CometBFT and the application.
+
+> `PrepareProposal` and `ProcessProposal` are understood as implemented functions. The easiest 
+> (and thus the canonical) way
+> to ensure the requirements is to make sure that `PrepareProposal` only proposes blocks v that
+> satisfy a (mathematical function) `valid(v, bc(h-1))`, and `ProcessProposal` just evaluates this
```

Comment:
This is a local coherence property, namely `ProcessProposal` at correct process $p$ must accept any possible output of `PrepareProposal` at the same process $p$.

Reply:
This "local coherence" can be put together with Requirement 5 and we have the properties we want.